README.md
# ibtcl

Interbase/Firebird extension for Tcl.

Recompiled and tested 04.11.2020 on Windows 10, Ubuntu 20.04, RHEL 7.4
                                           
Changed 04.11.2015, bugfix:

        - Fixed SQLDIALECT usage in the ib_open.
        - Support for FLOAT and INT64 db fields.
        - Makefile updated to build shared lib.

Changed 24.10.2015, version 0.1.1:

        - Optional database role and encoding for the ib_open call.
        - Uses UTF-8 conversions.
        - Defines SQLDIALECT in cmd.c (3 by default).
        - Defines TCLDATEFORMAT in cmd.c (string datetime instead of tcltime by default).
        - Can be compiled on Windows, MSVC makefile included.

Original from:

http://ftp.uni-hannover.de/tcl/mirror/ftp.procplace.com/sorted/packages-8.0/databases/ibtcl/1.0/ibtcl-01.tar.gz

Copyright (C) 1998 Oleg Checkulaev




changes.txt
original source at http://www.google.com.ua/search?q=ibtcl-01.tar.gz

modified by oleg oleinick, ooa64@ua.fm

was:	sql dialect 1
now:	sql dialect 3

was:	ib_open db user password
now:	ib_open db user password ?role ?tclencoding??

was:	DATE value from database is returned in seconds.
now:	DATE value from database is returned in string.



test.tcl
catch {load ibtcl011tg.dll}
catch {load ibtcl011t.dll}
catch {load ibtcl011.dll}
catch {load ./ibtcl.so}

set dbn "localhost:c:/FireBird/Data/test.fdb"
set dbh [ib_open $dbn SYSDBA masterkey]
set stm [ib_exec $dbh {select current_timestamp from RDB$DATABASE}]

puts isquery=[ib_isquery $stm]
puts fields=[ib_fields $stm]
puts field0=[ib_fieldname $stm 0]
puts fetch=[ib_fetch -n 1 $stm a]
parray a

ib_free_stmt $stm
ib_close $dbh


id.c
/*
* ibtcl
*
* Copyright (C) 1998 Oleg Checkulaev
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/

#include <stdlib.h>
#include <string.h>
#include "ibtclInt.h"

int ib_close_conn( Tcl_Interp *ip, IB_Connection *con );
void ib_close_stmt( IB_Statement *s );

static char const rcsid[] = "$Id: id.c,v 0.2 1998/10/01 14:18:49 coa Exp $";

void ib_set_conn_id( IB_ClientData *cd, IB_Connection *connid ) {
	Tcl_HashEntry *hent;
	int hnew;

	Tcl_InitHashTable( &(connid->stmt_hash), TCL_STRING_KEYS );

	hent = Tcl_CreateHashEntry( &(cd->db_hash), connid->id, &hnew );
	Tcl_SetHashValue( hent, (ClientData)connid );
}



IB_Connection* ib_get_conn_id( IB_ClientData *cd, char *id) {
	Tcl_HashEntry *hent;

	hent = Tcl_FindHashEntry( &(cd->db_hash), id );
	if( hent == NULL ) {
		return (IB_Connection*)NULL;
	}

	return (IB_Connection*) Tcl_GetHashValue( hent );
}



void ib_del_conn_id( IB_ClientData *cd, char *id ) {
	Tcl_HashEntry *hent, *hent2, *hent3;
	Tcl_HashSearch hsearch;
	IB_Connection *connid;
	IB_Statement *sid;

	hent = Tcl_FindHashEntry( &(cd->db_hash), id );
	if(hent == NULL) {
		return;
	}

	connid = (IB_Connection*) Tcl_GetHashValue( hent );

	hent2 = Tcl_FirstHashEntry( &(connid->stmt_hash), &hsearch );
	while( hent2 != NULL ) {
		sid = (IB_Statement*) Tcl_GetHashValue( hent2 );
		ib_close_stmt( sid );
		hent3 = Tcl_FindHashEntry( &(cd->stmt_hash), sid->id );
		if( hent3 != NULL ) {
			Tcl_DeleteHashEntry( hent3 );
		}
		ckfree( (char *)sid );
		hent2 = Tcl_NextHashEntry( &hsearch );
	}

	Tcl_DeleteHashTable( &(connid->stmt_hash) );
	Tcl_DeleteHashEntry( hent );

	return;
}



void ib_set_stmt_id( IB_ClientData *cd, IB_Connection *con, IB_Statement *st ) {
	Tcl_HashEntry *hent;
	int hnew;


	hent = Tcl_CreateHashEntry( &(cd->stmt_hash), st->id, &hnew);
	Tcl_SetHashValue( hent, (ClientData)st );

	if(con != NULL) {
		hent = Tcl_CreateHashEntry( &(con->stmt_hash), st->id, &hnew );
		Tcl_SetHashValue( hent, (ClientData)st );
	}
}



IB_Statement* ib_get_stmt_id( IB_ClientData *cd, char *id) {
	Tcl_HashEntry *hent;

	hent = Tcl_FindHashEntry( &(cd->stmt_hash), id );
	if( hent == NULL ) {
		return (IB_Statement*)NULL;
	}

	return (IB_Statement*) Tcl_GetHashValue( hent );
}



void ib_del_stmt_id( IB_ClientData *cd, char *id ) {
	Tcl_HashEntry	*hent;
	Tcl_HashEntry	*hent2;
	IB_Statement	*stmt;

	hent = Tcl_FindHashEntry( &(cd->stmt_hash), id );
	if( hent==NULL ) {
		return;
	}

	stmt = (IB_Statement*)Tcl_GetHashValue( hent );
	if( stmt->con != NULL ) {
		hent2 = Tcl_FindHashEntry( &(stmt->con->stmt_hash), id );
		if( hent2 != NULL ) {
			Tcl_DeleteHashEntry( hent2 );
		}
	}

	Tcl_DeleteHashEntry( hent );

	return;
}


ibtcl.h
#ifndef __IBTCL_H_COA__
#define __IBTCL_H_COA__
/*
* ibtcl
*
* Copyright (C) 1998 Oleg Checkulaev
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/
/*
$Id: ibtcl.h,v 0.2 1998/11/02 07:50:06 coa Exp coa $
*/

#include <tcl.h>

#undef TCL_STORAGE_CLASS
#define TCL_STORAGE_CLASS DLLEXPORT

EXTERN int Ibtcl_Init( Tcl_Interp *interp );
EXTERN int Ibtcl_SafeInit( Tcl_Interp *interp );

#endif


ibtclsh.c
/*
* ibtcl
*
* Copyright (C) 1998 Oleg Checkulaev
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/

#include <stdio.h>
#include "ibtcl.h"

static char const rcsid[] = "$Id: ibtclsh.c,v 0.1 1998/10/01 09:06:10 coa Exp $";

main( int argc, char **argv ) {
	Tcl_Main( argc, argv, Ibtcl_Init );
}


ibtclInt.h
#ifndef __IBTCLINT_H_COA__
#define __IBTCLINT_H_COA__
/*
* ibtcl
*
* Copyright (C) 1998 Oleg Checkulaev
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/
/*
$Id: ibtclInt.h,v 0.3 1998/10/09 11:33:12 coa Exp $
*/

#include <ibase.h>
#include "ibtcl.h"


typedef struct {
	Tcl_HashTable db_hash;
	Tcl_HashTable stmt_hash;
	int num_db;
	int num_stmt;
} IB_ClientData;

typedef struct {
	char id[32];
	isc_db_handle dbh;
	isc_tr_handle trh;
	Tcl_HashTable stmt_hash;
	Tcl_Encoding enc;
	Tcl_DString enc_buf;
} IB_Connection;

typedef struct {
	char id[32];
	int f_maxlen; /* max length of fields */
	isc_stmt_handle stmth;
	IB_Connection *con;
	XSQLDA *xd;
} IB_Statement;


/* ibtcl functions themselves */
int do_ib_test( ClientData, Tcl_Interp*, int, char** );
int do_ib_open( ClientData, Tcl_Interp*, int, char** );
int do_ib_close( ClientData, Tcl_Interp*, int, char** );
int do_ib_exec( ClientData, Tcl_Interp*, int, char** );
int do_ib_free_stmt( ClientData, Tcl_Interp*, int, char** );
int do_ib_fetch( ClientData, Tcl_Interp*, int, char** );
int do_ib_fetch2proc( ClientData, Tcl_Interp*, int, char** );
int do_ib_skip( ClientData, Tcl_Interp*, int, char** );
int do_ib_fieldname( ClientData, Tcl_Interp*, int, char** );
int do_ib_fields( ClientData, Tcl_Interp*, int, char** );
int do_ib_isquery( ClientData, Tcl_Interp*, int, char** );

/* Id manipulation functions */
void ib_set_conn_id( IB_ClientData*, IB_Connection* );
IB_Connection* ib_get_conn_id( IB_ClientData*, char* );
void ib_del_conn_id( IB_ClientData*, char* );
void ib_set_stmt_id( IB_ClientData*, IB_Connection*, IB_Statement* );
IB_Statement* ib_get_stmt_id( IB_ClientData*, char* );
void id_del_stmt_id( IB_ClientData, char );

#endif


ibtcl.c
/*
* ibtcl
*
* Copyright (C) 1998 Oleg Checkulaev
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/

#include <stdlib.h>
#include "ibtclInt.h"

static char const rcsid[] = "$Id: ibtcl.c,v 0.3 1998/11/02 07:50:06 coa Exp coa $";

int ib_close_conn( Tcl_Interp *ip, IB_Connection *con );

static void ibtcl_AtExit( ClientData cData ) {
	IB_ClientData *cd = (IB_ClientData *)cData;
	Tcl_HashEntry *hent;
	Tcl_HashSearch hsearch;
	IB_Connection *con;

	while( (hent=Tcl_FirstHashEntry(&(cd->db_hash),&hsearch)) != NULL ) {
		con = (IB_Connection*)Tcl_GetHashValue( hent );
		ib_del_conn_id( cd, con->id );
		ib_close_conn( NULL, con );
	}

	Tcl_DeleteHashTable( &(cd->db_hash) );
	Tcl_DeleteHashTable( &(cd->stmt_hash) );

	Tcl_DeleteExitHandler( ibtcl_AtExit, cData);
}


static void ibtcl_shutdown( ClientData cData, Tcl_Interp *interp ) {
	ibtcl_AtExit( cData );
}



EXTERN int Ibtcl_Init (Tcl_Interp *interp) {
	IB_ClientData *cd;

#ifdef USE_TCL_STUBS
	if (Tcl_InitStubs(interp, "8.1", 0) == 0L) {
		return TCL_ERROR;
	}
#endif

	cd = (IB_ClientData*) ckalloc( sizeof(IB_ClientData) );
	Tcl_InitHashTable( &(cd->db_hash), TCL_STRING_KEYS );
	Tcl_InitHashTable( &(cd->stmt_hash), TCL_STRING_KEYS );
	cd->num_db = 0;
	cd->num_stmt = 0;

	Tcl_CallWhenDeleted( interp, ibtcl_shutdown, (ClientData)cd );
	Tcl_CreateExitHandler( ibtcl_AtExit, (ClientData)cd );

	Tcl_CreateCommand(interp, "ib_test", (Tcl_CmdProc *)do_ib_test, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);

	Tcl_CreateCommand(interp, "ib_open", (Tcl_CmdProc *)do_ib_open, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_close", (Tcl_CmdProc *)do_ib_close, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);

	Tcl_CreateCommand(interp, "ib_exec", (Tcl_CmdProc *)do_ib_exec, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_free_stmt", (Tcl_CmdProc *)do_ib_free_stmt, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);

	Tcl_CreateCommand(interp, "ib_fetch", (Tcl_CmdProc *)do_ib_fetch, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_fetch2proc", (Tcl_CmdProc *)do_ib_fetch2proc, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_skip", (Tcl_CmdProc *)do_ib_skip, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_fieldname", (Tcl_CmdProc *)do_ib_fieldname, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);
	Tcl_CreateCommand(interp, "ib_fields", (Tcl_CmdProc *)do_ib_fields, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);

	Tcl_CreateCommand(interp, "ib_isquery", (Tcl_CmdProc *)do_ib_isquery, (ClientData)cd, (Tcl_CmdDeleteProc*)NULL);

	Tcl_PkgProvide( interp, "ibtcl", "0.1" );
	
	return TCL_OK;
}


EXTERN int Ibtcl_SafeInit( Tcl_Interp *ip ) {
	return Ibtcl_Init( ip );
}
